### PR TITLE
Migrate Restart, FIPS, Proxy, LVM and SSL cert tests

### PR DESCRIPTION
**Pull request by movence**
# Description of changes
- Migrate Restart, FIPS, LVM , Proxy and SSL Cert tests
- Move acceptance test to ec2 matrix
- Add target OS param to test matrices
- Add proxy resource to ec2 main.tf to support proxy test

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
Run in private github repo

### CODE REVIEW

File: test/proxy/proxy_unix.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+//go:build !windows
+
+package proxy
+
+const commonConfigPath = "/opt/aws/amazon-cloudwatch-agent/etc/common-config.toml"
+
+func GetCommandToCreateProxyConfig(proxyUrl string) []string {
+	return []string{
+		"echo [proxy] | sudo tee -a /opt/aws/amazon-cloudwatch-agent/etc/common-config.toml",
```
Comment (movence) [inline test/proxy/proxy_unix.go:12]: this is ugly but bash subshell kept throwing errors with heredoc.

File: test/restart/agent_config.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (movence) [inline test/restart/agent_config.json:1]: this is a placeholder agent config that common module expects. It is not used to run the agent.

File: terraform/ec2/linux/main.tf [hunk 3]
```diff
@@ -68,7 +132,7 @@ resource "null_resource" "integration_test" {
   provisioner "remote-exec" {
     inline = [
       "echo sha ${var.cwa_github_sha}",
-      "cloud-init status --wait",
+      "sudo cloud-init status --wait",
```
Comment (sethAmazon) [inline terraform/ec2/linux/main.tf:135]: Does this need super user?

File: test/lvm/lvm_test.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+func (t LVMTestRunner) GetAgentConfigFileName() string {
+	return "config.json"
+}
+
+func (t LVMTestRunner) GetMeasuredMetrics() []string {
+	return []string{"disk_free", "disk_used", "disk_total"}
+}
+
+func (t *LVMTestRunner) SetupBeforeAgentRun() error {
```
Comment (sethAmazon) [inline test/lvm/lvm_test.go:93]: Does this need to also set up the config for the agent now?

File: test/fips/fips_test.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	agentConfigPath      = "/opt/aws/amazon-cloudwatch-agent/bin/config.json"
+)
+
+var envMetaDataStrings = &(environment.MetaDataStrings{})
+
+func init() {
+	environment.RegisterEnvironmentMetaDataFlags(envMetaDataStrings)
+}
+
+func TestFIPS(t *testing.T) {
```
Comment (sethAmazon) [inline test/fips/fips_test.go:33]: Why are these in the style of the test suite but are not in the metric_value_benchmark test or any newly created test suite?

File: test/lvm/lvm_test.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	agentRuntime = 3 * time.Minute
+)
+
+var envMetaDataStrings = &(environment.MetaDataStrings{})
+
+func init() {
+	environment.RegisterEnvironmentMetaDataFlags(envMetaDataStrings)
+}
+
+type LVMTestRunner struct {
```
Comment (sethAmazon) [inline test/lvm/lvm_test.go:31]: Why are these in the style of the test suite but we only have one test per suite. I feel like the test should stand on its own or we combine multiple into a single suite.

Comment (movence) [reply test/lvm/lvm_test.go:31]: It's not in the test suite style but using `base_test_runner` which is used by a suite. This is to have the same test runner structure for better readability.

File: test/lvm/lvm_test.go (hunk 4 above)
Comment (movence) [reply test/lvm/lvm_test.go:93]: It is done by another function `runAgent`

File: terraform/ec2/linux/main.tf (hunk 3 above)
Comment (movence) [reply terraform/ec2/linux/main.tf:135]: it was throwing permission error:
```
Traceback (most recent call last):
null_resource.integration_test_setup (remote-exec):   File "/usr/bin/cloud-init", line 11, in <module>
null_resource.integration_test_setup (remote-exec):     load_entry_point('cloud-init==19.3', 'console_scripts', 'cloud-init')()
null_resource.integration_test_setup (remote-exec):   File "/usr/lib/python2.7/site-packages/cloudinit/cmd/main.py", line 893, in main
null_resource.integration_test_setup (remote-exec):     get_uptime=True, func=functor, args=(name, args))
null_resource.integration_test_setup (remote-exec):   File "/usr/lib/python2.7/site-packages/cloudinit/util.py", line 2564, in log_time
null_resource.integration_test_setup (remote-exec):     ret = func(*args, **kwargs)
null_resource.integration_test_setup (remote-exec):   File "/usr/lib/python2.7/site-packages/cloudinit/cmd/status.py", line 53, in handle_status_args
null_resource.integration_test_setup (remote-exec):     init.read_cfg()
null_resource.integration_test_setup (remote-exec):   File "/usr/lib/python2.7/site-packages/cloudinit/stages.py", line 175, in read_cfg
null_resource.integration_test_setup (remote-exec):     self._cfg = self._read_cfg(extra_fns)
null_resource.integration_test_setup (remote-exec):   File "/usr/lib/python2.7/site-packages/cloudinit/stages.py", line 184, in _read_cfg
null_resource.integration_test_setup (remote-exec):     return merger.cfg
null_resource.integration_test_setup (remote-exec):   File "/usr/lib/python2.7/site-packages/cloudinit/helpers.py", line 278, in cfg
null_resource.integration_test_setup (remote-exec):     self._cfg = self._read_cfg()
null_resource.integration_test_setup (remote-exec):   File "/usr/lib/python2.7/site-packages/cloudinit/helpers.py", line 268, in _read_cfg
null_resource.integration_test_setup (remote-exec):     cfgs.extend(self._get_instance_configs())
null_resource.integration_test_setup (remote-exec):   File "/usr/lib/python2.7/site-packages/cloudinit/helpers.py", line 242, in _get_instance_configs
```

File: test/ssl_cert/ssl_cert_test.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+//go:build !windows
+
+package acceptance
```
Comment (SaxyPandaBear) [inline test/ssl_cert/ssl_cert_test.go:5]: nit - wrong package name

File: test/restart/restart_test.go [hunk 8]
```diff
@@ -0,0 +1,17 @@
+package restart
```
Comment (SaxyPandaBear) [inline test/restart/restart_test.go:1]: nit - missing license header

File: test/restart/common.go [hunk 9]
```diff
@@ -0,0 +1,30 @@
+package restart
```
Comment (SaxyPandaBear) [inline test/restart/common.go:1]: nit - missing license header

File: test/lvm/lvm_test.go [hunk 10]
```diff
@@ -0,0 +1,119 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+//go:build !windows
+
+package acceptance
```
Comment (SaxyPandaBear) [inline test/lvm/lvm_test.go:5]: nit - package name

File: test/fips/fips_test.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+//go:build !windows
+
+package acceptance
```
Comment (SaxyPandaBear) [inline test/fips/fips_test.go:5]: nit - package name

File: validator/main.go [hunk 12]
```diff
@@ -8,38 +8,53 @@ import (
...
-			log.Fatalf("Prepare for validation failed: %v \n", err)
+	// validator calls test code to get around OOM issue on windows hosts while running go test
+	if len(*configPath) == 0 && len(*runTest) > 0 {
+		// execute test without parsing or processing configuration yaml
+		if strings.Contains(*runTest, "restart") {
+			err := restart.LogCheck("\"{& Function countLogLines { Param ( [Parameter(Mandatory = $true)] [string]$log_path ) if (Test-Path -LiteralPath '$log_path') { return (Get-Content $log_path).Length } else { return '0' } } $cwa_log=countLoglines('$Env:ProgramData\\Amazon\\AmazonCloudWatchAgent\\Logs\\amazon-cloudwatch-agent.log') Write-Output 'cwa_log:$cwa_log' }\"")
+			if len(err) > 0 {
+				log.Fatalf("Failed to validate restart: %v", err)
+			}
+		}
```
[the remark this answers is not in the document]
Comment (movence) [reply validator/main.go:40]: Yeah, good point. I failed cover other platforms after adding that new `test-name` param only to Windows. Will address it.

File: generator/test_case_generator.go [hunk 13]
```diff
@@ -123,14 +131,18 @@ func genMatrix(testType string, testConfigs []testConfig) []matrixRow {
...
+		targetOSes := strings.Join(testConfig.targetOs, ",")
+		for _, test := range testMatrix {
 			row := matrixRow{TestDir: testConfig.testDir, TestType: testType, TerraformDir: testConfig.terraformDir}
 			err = mapstructure.Decode(test, &row)
 			if err != nil {
 				log.Panicf("can't decode map test %v to metric line struct with error %v", testConfig, err)
 			}
-			testMatrixComplete = append(testMatrixComplete, row)
+
+			if len(testConfig.targetOs) == 0 || strings.Contains(targetOSes, row.Os) {
```
[the remark this answers is not in the document]
Comment (movence) [reply generator/test_case_generator.go:143]: Agree. The idea is to add a `row` to test matrix when either `targetOs` is `nil` or `targetOs` contains the current os.

File: generator/test_case_generator.go [hunk 14]
```diff
@@ -36,65 +36,72 @@ type testConfig struct {
...
+	// target specific OS(es) so that only limited targets are tested. doesn't apply to containers
+	// e.g. ["rhel8"] or ["rhel8", "ubuntu-20.04"]
+	targetOs map[string]struct{}
 }
 
 // you can't have a const map in golang
 var testTypeToTestConfig = map[string][]testConfig{
 	"ec2_gpu": {
-		{"./test/nvidia_gpu", ""},
+		{"./test/nvidia_gpu", "", nil},
```
Comment (SaxyPandaBear) [inline generator/test_case_generator.go:47]: we don't need to specify `nil` here because that's the default value of a map

File: generator/test_case_generator.go [hunk 15]
```diff
@@ -123,18 +130,33 @@ func genMatrix(testType string, testConfigs []testConfig) []matrixRow {
...
+
+func shouldAddTest(os string, targets map[string]struct{}) bool {
+	if targets == nil {
+		return true
+	}
+	if _, ok := targets[os]; ok {
+		return true
+	}
+
+	return false
```
Comment (SaxyPandaBear) [inline generator/test_case_generator.go:157]: This could be simplified to 
```go
_, ok := targets[os]
return ok
```

**Review by SaxyPandaBear — APPROVED** [review]

File: test/test_runner/base_test_runner.go [hunk 16]
```diff
@@ -101,9 +101,23 @@ func (t *BaseTestRunner) SetAgentConfig(agentConfig AgentConfig) {
...
+		testGroupResult = t.TestRunner.Validate()
+	}
+	if testGroupResult.GetStatus() != status.SUCCESSFUL {
+		log.Printf("%v test group failed due to %v", testName, err)
+	}
+
+	return testGroupResult
+}
+
+func (t *TestRunner) RunAgent() (status.TestGroupResult, error) {
```
Comment (zhihonl) [inline test/test_runner/base_test_runner.go:118]: We want the behavior to be that we have a default RunAgent method for all test_runner, and then each test_runner can implement their own version if provided.

Is there a test run to show that this is working as intended?

Comment (movence) [reply test/test_runner/base_test_runner.go:118]: yeah, `RunAgent` gets executed by the default `Run` and each test_runner can override it as needed

Comment (movence) [reply test/test_runner/base_test_runner.go:118]: This was ran from my personal fork https://github.com/movence/amazon-cloudwatch-agent/actions/runs/5158969625/jobs/9293404442
```
null_resource.integration_test_run (remote-exec): === RUN   TestUserdata
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:09 Testing all EC2 plugins
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:09 Running UserdataTest
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:09 Starting agent using agent config file agent_configs/config.json
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:09 Copy File agent_configs/config.json to /opt/aws/amazon-cloudwatch-agent/bin/config.json
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:09 File agent_configs/config.json abs path /home/ec2-user/amazon-cloudwatch-agent-test/test/userdata/agent_configs/config.json
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:10 File : agent_configs/config.json copied to : /opt/aws/amazon-cloudwatch-agent/bin/config.json
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:10 Agent has started
null_resource.integration_test_run: Still creating... [3m20s elapsed]
null_resource.integration_test_run: Still creating... [3m30s elapsed]
null_resource.integration_test_run: Still creating... [3m40s elapsed]
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:40 Agent has been running for : 30s
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 Agent is stopped
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 Delete file /opt/aws/amazon-cloudwatch-agent/bin/config.json
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 Removed file: /opt/aws/amazon-cloudwatch-agent/bin/config.json
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 instruction {InstanceId {<nil>}} provider HostDimensionProvider returned dimension {<nil> <nil> {}}
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 instruction {InstanceId {<nil>}} provider LocalInstanceIdDimensionProvider returned dimension {0xc000038220 0xc000038230 {}}
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 Result dim is : InstanceId, i-0544dd42c6d9882a7
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 instruction {cpu {0xc000159c30}} provider HostDimensionProvider returned dimension {<nil> <nil> {}}
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 instruction {cpu {0xc000159c30}} provider LocalInstanceIdDimensionProvider returned dimension {<nil> <nil> {}}
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 instruction {cpu {0xc000159c30}} provider LocalImageIdDimensionProvider returned dimension {<nil> <nil> {}}
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 instruction {cpu {0xc000159c30}} provider LocalInstanceTypeDimensionProvider returned dimension {<nil> <nil> {}}
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 instruction {cpu {0xc000159c30}} provider CustomDimensionProvider returned dimension {0xc0000382b0 0xc0000382c0 {}}
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 Result dim is : cpu, cpu-total
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 Metric query input dimensions
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 	Dimensions:
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 		Dim(name="InstanceId", val="i-0544dd42c6d[988](https://github.com/movence/amazon-cloudwatch-agent/actions/runs/5158969625/jobs/9293404442#step:7:989)2a7"
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 		Dim(name="cpu", val="cpu-total"
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 Metric data input: namespace UserdataTest, name cpu_time_active_userdata, stat Average, period 10
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 Metric values are : [361.73 361.08]
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 metric values are [361.73 361.08]
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 The average value 361.405000 for metric cpu_time_active_userdata are within bound [0.000000, 0.000000]
null_resource.integration_test_run (remote-exec): --- PASS: TestUserdata (31.34s)
```

Comment (zhihonl) [reply test/test_runner/base_test_runner.go:118]: https://github.com/movence/amazon-cloudwatch-agent/actions/runs/5158969625/jobs/9293404442#step:7:970
I don't think it is. Userdata test implements `RunAgent` method so that it does nothing about starting and configuring the agent since those should be done using EC2 userdata. In your test run it's running the default RunAgent method.

Comment (movence) [reply test/test_runner/base_test_runner.go:118]: You are right. I should have overlooked the logs. I changed the userdata to have its own `Run` method, so it calls its own function not the one in `base_test_runner`.

```
null_resource.integration_test (remote-exec): === RUN   TestUserdata
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 Testing all EC2 plugins
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 Running UserdataTest
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 Running test for runAgent in userdata mode
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 instruction {InstanceId {<nil>}} provider HostDimensionProvider returned dimension {<nil> <nil> {}}
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 instruction {InstanceId {<nil>}} provider LocalInstanceIdDimensionProvider returned dimension {0xc000270070 0xc000270080 {}}
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 Result dim is : InstanceId, i-071bc4321aa09406d
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 instruction {cpu {0xc0001d3b20}} provider HostDimensionProvider returned dimension {<nil> <nil> {}}
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 instruction {cpu {0xc0001d3b20}} provider LocalInstanceIdDimensionProvider returned dimension {<nil> <nil> {}}
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 instruction {cpu {0xc0001d3b20}} provider LocalImageIdDimensionProvider returned dimension {<nil> <nil> {}}
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 instruction {cpu {0xc0001d3b20}} provider LocalInstanceTypeDimensionProvider returned dimension {<nil> <nil> {}}
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 instruction {cpu {0xc0001d3b20}} provider CustomDimensionProvider returned dimension {0xc000270100 0xc000270110 {}}
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 Result dim is : cpu, cpu-total
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 Metric query input dimensions
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 	Dimensions:
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 		Dim(name="InstanceId", val="i-071bc[432](https://github.com/movence/amazon-cloudwatch-agent/actions/runs/5164785471/jobs/9303925508#step:6:433)1aa09406d"
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 		Dim(name="cpu", val="cpu-total"
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 Metric data input: namespace UserdataTest, name cpu_time_active_userdata, stat Average, period 10
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 Metric values are : [125.72000000000003 110.20000000000002]
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 metric values are [125.72000000000003 110.20000000000002]
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 The average value 117.960000 for metric cpu_time_active_userdata are within bound [0.000000, 0.000000]
null_resource.integration_test (remote-exec): --- PASS: TestUserdata (0.06s)
```

File: test/userdata/userdata_test.go [hunk 17]
```diff
@@ -96,9 +75,70 @@ func (t *UserdataTestRunner) validateCpuMetric(metricName string) status.TestRes
...
+
+	testGroupResult = t.Validate()
+	if testGroupResult.GetStatus() != status.SUCCESSFUL {
+		log.Printf("%v test run failed", testName)
+	}
+
+	return testGroupResult
+}
+
+func (t UserdataTestRunner) RunAgent() (status.TestGroupResult, error) {
```
[the remark this answers is not in the document]
Comment (movence) [reply test/userdata/userdata_test.go]: yeah, good catch. removed

**Comment by sethAmazon** [comment]
Fix windows compile

File: terraform/ec2/linux/main.tf [hunk 18]
```diff
@@ -78,6 +142,53 @@ resource "null_resource" "integration_test" {
...
+
+  depends_on = [
+    null_resource.integration_test_setup,
+  ]
+}
+
+resource "null_resource" "integration_test_wait" {
+  count = contains(local.reboot_required_tests, var.test_dir) ? 1 : 0
+  provisioner "local-exec" {
+    command = "echo Sleeping for 3m after initiating instance restart && sleep 180"
```
Comment (sethAmazon) [inline terraform/ec2/linux/main.tf:177]: Do we really need to w8 3 minutes. Any chance we can do connect retry with wait times instead?

File: generator/test_case_generator.go (hunk 14 above)
Comment (SaxyPandaBear) [reply generator/test_case_generator.go:47]: Huh. [TIL](https://go.dev/play/p/XOgSO3B3mkq). Okay, well then personally I would rather put names on the struct values, rather than use positional arguments, but this is non blocking.

**Review by SaxyPandaBear — APPROVED** [review]

**Review by zhihonl — APPROVED** [review]

**Comment by movence** [comment]
We should visit for go package being excluded for windows build and make sure CWA features on Windows are also getting tested correctly.

**Review by zhihonl — APPROVED** [review]
